From 9922dc49e163a01ea082bc9cc9975deb51e904ca Mon Sep 17 00:00:00 2001
From: Etienne Carriere <etienne.carriere@foss.st.com>
Date: Tue, 2 Jul 2024 10:25:56 +0200
Subject: [PATCH 1086/1695] drivers: clk: restore clock on
 stm32_gate_ready_endisable() failure

Return to previous state of the clock in case stm32_gate_ready_endisable() failed.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git a6ff9611cdcdeff4b4c8b2c04caa8ce59ed0b96e drivers: clk: restore clock on stm32_gate_ready_endisable() failure]
Change-Id: I9468e6117a44b9e1d7078b628c5d01f8f162e8d2
Signed-off-by: Etienne Carriere <etienne.carriere@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/390714
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Tested-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
Domain-Review: Patrick DELAUNAY <patrick.delaunay@foss.st.com>
Reviewed-by: Thomas BOURGOIN <thomas.bourgoin@foss.st.com>
---
 core/drivers/clk/clk-stm32-core.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/core/drivers/clk/clk-stm32-core.c b/core/drivers/clk/clk-stm32-core.c
index d2a716d83..4d84f9d86 100644
--- a/core/drivers/clk/clk-stm32-core.c
+++ b/core/drivers/clk/clk-stm32-core.c
@@ -144,10 +144,20 @@ TEE_Result stm32_gate_wait_ready(uint16_t gate_id, bool ready_on)
 static TEE_Result stm32_gate_ready_endisable(uint16_t gate_id, bool enable,
 					     bool wait_rdy)
 {
+	TEE_Result res = TEE_ERROR_GENERIC;
+
 	stm32_gate_endisable(gate_id, enable);
 
-	if (wait_rdy)
-		return stm32_gate_wait_ready(gate_id + 1, enable);
+	if (wait_rdy) {
+		res = stm32_gate_wait_ready(gate_id + 1, enable);
+		if (res) {
+			stm32_gate_endisable(gate_id, !enable);
+			if (stm32_gate_wait_ready(gate_id + 1, !enable))
+				panic("Gate failed to sync");
+
+			return res;
+		}
+	}
 
 	return TEE_SUCCESS;
 }
-- 
2.47.2

